core/dbt/contracts/graph/model_config.py

```diff
@@ -460,6 +461,20 @@ class NodeConfig(NodeAndTestConfig):
     grants: Dict[str, Any] = field(
         default_factory=dict, metadata=MergeBehavior.DictKeyAppend.meta()
     )
+    docs: Dict[str, Any] = field(
+        default_factory=dict,
+        metadata=MergeBehavior.Update.meta(),
+    )
```

Comment:
Indeed, what we really want here is to merge the values of `node.docs` and `node.config.docs`, just as we do today with `node.meta` and `node.config.meta`. If we can do that, the change will be both sensible and backwards compatible.

There are a few relevant spots for that (and I'm probably not catching all of them, so @dbt-labs/core-language should really advise):

https://github.com/dbt-labs/dbt-core/blob/d5608dca322c63f1a894174b1f280ae55a217ab8/core/dbt/parser/base.py#L284-L287

https://github.com/dbt-labs/dbt-core/blob/d5608dca322c63f1a894174b1f280ae55a217ab8/core/dbt/parser/schemas.py#L807-L821

We're moving toward a world where:
- `config` can accept any values, set/inherit them via config hierarchy, and get them for use at runtime
- any `config` values with important metadata implications should move up to top-level node properties (`tags`, `meta`, etc today; probably `materialized`, too)

Reply:
Thanks @jtcohen6 , I will have a look.

Reply:
Thanks for the pointers @jtcohen6 

We actually also need to remove/modify this part handling the "legacy" docs (I spent way too much time figuring out why `node.docs` was not getting populated): https://github.com/dbt-labs/dbt-core/blob/d63ad4cd82619dafed7d68194df08d256c15f85b/core/dbt/contracts/graph/parsed.py#L524

Is the goal to make the docs dict attached to a dataclass with `show`/`node_color` or would we want it to be just a generic `dict` like `meta`?

I will update the PR soon with my progress.

Reply:
> Is the goal to make the docs dict attached to a dataclass with `show`/`node_color` or would we want it to be just a generic `dict` like `meta`?

I think the former. `meta` should remain a sorta special case in its acceptance of _absolutely anything_, since dbt makes zero promises as far as validating its contents.

Of course, when `docs` is serialized into `manifest.json`, it should just be a standard dictionary with up to 2 keys, `show` (always set) and `node_color` (optionally set).

Reply:
I have a branch working locally when making `docs` a `dict` but making it a dataclass is more complex and raises a few errors.

1. Making `node_color` optional still requires setting a Default value, so we might need to have some specific code to handle the deletion of the attribute if `node_color = None`. The error I get without setting a default value is `TypeError: non-default argument 'node_color' follows default argument`
2. When adding the code to populate `docs` from `config.docs` with the following snippets, mahsumaro raises the error `NameError: name 'dbt' is not defined` which I can't pinpoint why. If I remove the new lines for "docs", `docs` is not moved from across from `config` (expected) but it also doesn't raise the error.

Code in `core/dbt/parser/base.py`:
```
       # the original code for "meta"
        if "meta" in config_dict and config_dict["meta"]:
            parsed_node.meta = config_dict["meta"]

       # the new code for "docs"
        if "docs" in config_dict and config_dict["docs"]:
            parsed_node.docs = config_dict["docs"]
```

Debug error:
```
name 'dbt' is not defined
17:25:10.583468 [debug] [MainThread]: Traceback (most recent call last):
  File "/Users/bper/dev/dbt-core2/core/dbt/main.py", line 129, in main
    results, succeeded = handle_and_check(args)
  File "/Users/bper/dev/dbt-core2/core/dbt/main.py", line 191, in handle_and_check
    task, res = run_from_args(parsed)
  File "/Users/bper/dev/dbt-core2/core/dbt/main.py", line 238, in run_from_args
    results = task.run()
  File "/Users/bper/dev/dbt-core2/core/dbt/task/runnable.py", line 451, in run
    self._runtime_initialize()
  File "/Users/bper/dev/dbt-core2/core/dbt/task/runnable.py", line 159, in _runtime_initialize
    super()._runtime_initialize()
  File "/Users/bper/dev/dbt-core2/core/dbt/task/runnable.py", line 92, in _runtime_initialize
    self.load_manifest()
  File "/Users/bper/dev/dbt-core2/core/dbt/task/runnable.py", line 81, in load_manifest
    self.manifest = ManifestLoader.get_full_manifest(self.config)
  File "/Users/bper/dev/dbt-core2/core/dbt/parser/manifest.py", line 218, in get_full_manifest
    manifest = loader.load()
  File "/Users/bper/dev/dbt-core2/core/dbt/parser/manifest.py", line 403, in load
    self.write_manifest_for_partial_parse()
  File "/Users/bper/dev/dbt-core2/core/dbt/parser/manifest.py", line 549, in write_manifest_for_partial_parse
    manifest_msgpack = self.manifest.to_msgpack()
  File "/Users/bper/.venv/dbt-core2/lib/python3.9/site-packages/mashumaro/serializer/msgpack.py", line 40, in to_msgpack
    self.to_dict(**dict(DEFAULT_DICT_PARAMS, **dict_params)),
  File "<string>", line 9, in to_dict
  File "<string>", line 9, in <dictcomp>
  File "<string>", line 74, in __pack_union_Manifest_nodes__a52031cbc3504db8b4c0b1b1952cd791
NameError: name 'dbt' is not defined
```